src/lib.rs

```diff
@@ -99,60 +99,40 @@ macro_rules! macro_attr {
 macro_rules! macro_attr_impl {
     (
         $(#[$($attrs:tt)+])*
-        enum $($it:tt)+
+        $(pub $(($($vis:tt)+))?)? enum $($it:tt)+
     ) => {
         $crate::macro_attr_impl! {
-            @split_attrs [enum $($it)+]
+            @split_attrs [$(pub $(($($vis)+))?)? enum $($it)+]
             [] []
             [$([$($attrs)+])*]
         }
     };
     (
         $(#[$($attrs:tt)+])*
-        pub $(($($vis:tt)+))? enum $($it:tt)+
+        $(pub $(($($vis:tt)+))?)? struct $($it:tt)+
     ) => {
         $crate::macro_attr_impl! {
-            @split_attrs [pub $(($($vis)+))? enum $($it)+]
+            @split_attrs [$(pub $(($($vis)+))?)? struct $($it)+]
             [] []
             [$([$($attrs)+])*]
         }
     };
     (
         $(#[$($attrs:tt)+])*
-        struct $($it:tt)+
+        $(pub $(($($vis:tt)+))?)? trait $($it:tt)+
     ) => {
         $crate::macro_attr_impl! {
-            @split_attrs [struct $($it)+]
-            [] []
-            [$([$($attrs)+])*]
-        }
-    };
-    (
-        $(#[$($attrs:tt)+])*
-        pub $(($($vis:tt)+))? struct $($it:tt)+
-    ) => {
-        $crate::macro_attr_impl! {
-            @split_attrs [pub $(($($vis)+))? struct $($it)+]
-            [] []
-            [$([$($attrs)+])*]
-        }
-    };
-    (
-        $(#[$($attrs:tt)+])*
-        trait $($it:tt)+
-    ) => {
-        $crate::macro_attr_impl! {
-            @split_attrs [trait $($it)+]
+            @split_attrs [pub $(($($vis)+))? trait $($it)+]
             [] []
             [$([$($attrs)+])*]
         }
     };
     (
         $(#[$($attrs:tt)+])*
-        pub $(($($vis:tt)+))? trait $($it:tt)+
+        $vis:vis $ty:ident $($it:tt)+
```

Comment:
Could we use `$ty:ident` approach in "tt" branch also? I like this.

Reply:
Unfortunately no, it causes to ambiguity.